server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -43,8 +52,44 @@ public DesiredBalanceShardsAllocator(
             protected void processInput(DesiredBalanceInput desiredBalanceInput) {
                 if (desiredBalanceService.updateDesiredBalanceAndReroute(desiredBalanceInput, this::isFresh)) {
                     pendingReroute = true;
-                    rerouteServiceSupplier.get()
-                        .reroute("desired balance changed", Priority.NORMAL, ActionListener.wrap(() -> pendingReroute = false));
+                    boolean isFreshInput = isFresh(desiredBalanceInput);
+                    var listener = new ActionListener<ClusterState>() {
+                        @Override
+                        public void onResponse(ClusterState clusterState) {
+                            if (isFreshInput) {
+                                pendingReroute = false;
+                                triggerProcessedListeners(desiredBalanceInput.index());
+                            }
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            //TODO check exception type?
+                            triggerAllListenersFailure(e);
+                        }
+                    };
+                    rerouteServiceSupplier.get().reroute("desired balance changed", Priority.NORMAL, listener);
+                }
+            }
+
+            private void triggerProcessedListeners(long index) {
+                synchronized (pendingListeners) {
+                    while (true) {
+                        var trigger = pendingListeners.peekFirst();
+                        if (trigger == null || trigger.index > index) {
+                            break;
+                        }
+                        pendingListeners.pollFirst().listener.onResponse(null);
+                    }
+                }
+            }
+
+            private void triggerAllListenersFailure(Exception e) {
+                synchronized (pendingListeners) {
+                    DesiredBalancesListener listener;
+                    while ((listener = pendingListeners.pollLast()) != null) {
```

Comment:
Why complete them in reverse order here? I think we can use a regular `Queue<>` rather than a `Deque<>` if we don't need this.